### PR TITLE
Improve YOLOv5su mAP@0.5 on COCO128 while keeping inference under 200ms (exp-008)

### PR DESCRIPTION
**Pull request by coderXcode**
## Objective

Improve YOLOv5su mAP@0.5 on COCO128 while keeping inference under 200ms

## Current baseline

- Commit: `a1c4d7b587fabf892ddbb89b35f33fc7c9854a87`
- map50: 0.8289805608299404
- latency_ms: 32.61
- map50_95: 0.6588
- training_time_s: 598.0
- Environment: venv

## Motivating papers

- **arxiv:2211.07635** — Dennis Melamed, Karnik Ram, Vivek Roy, Kris Kitani. *Learnable Spatio-Temporal Map Embeddings for Deep Inertial Localization* (2022). <https://arxiv.org/pdf/2211.07635v1>

## Experiments attempted

| Experiment | Title | Status | Result |
|---|---|---|---|
| exp-007 | AdamW Optimizer with Tuned Learning Rate | rejected | map50=0.363467514918614 |
| exp-008 | Extended Fine-Tuning Epochs with Adjusted Warmup | implementation_ready | map50=0.8379805562463072 |
| exp-009 | Reduced Mosaic Augmentation for Precise Location Cues | rejected | map50=0.8289805608299404 |

## Rejected approaches

- **exp-007** (AdamW Optimizer with Tuned Learning Rate): primary metric did not improve vs baseline (0.8289805608299404 -> 0.363467514918614)
- **exp-009** (Reduced Mosaic Augmentation for Precise Location Cues): primary metric did not improve vs baseline (0.8289805608299404 -> 0.8289805608299404)

## Validated metrics

- map50: 0.8289805608299404 → mean 0.838 across 5 validation run(s) (stdev 0)
- Pre-ship confirmation: map50=0.8379805562463072

## Constraints

No hard constraints in the contract.

## Changed files

- `src/config.py`

## Reproduction

```bash
researchforge baseline run
researchforge experiment plan hyp-003
researchforge experiment import .researchforge/experiments/plan.yaml
researchforge experiment approve plan-003 --yes
researchforge experiment run plan-003
researchforge validate <run-id>
```

Evaluation command: `python benchmarks/evaluate.py` (result file: `artifacts/results.json`)

## Risks and limitations

- Requires short fine-tuning on COCO128 to adapt spatial embedding weights.
- Results were measured on a single machine (macOS-26.5.2-arm64-arm-64bit-Mach-O) in venv mode; they may not generalize beyond the tested conditions.
- venv execution isolates dependencies, not the machine — see the ResearchForge docs for the trust model.
- No test command is configured in the contract; no new tests were authored — test authoring is a Claude-assisted step (Phase 1E).

## ResearchForge report

`/Users/manasjoshi/Desktop/forge_hub/test_rf/yolov5/.researchforge/reports/engineering-report.md`

---

Draft PR generated by ResearchForge from recorded experiment data.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds a configurable COCO128 training, validation, and latency evaluation workflow for YOLOv5su, using a 15-epoch fine-tuning configuration and writing structured results to `artifacts/results.json`.

### 📊 Key Changes
- Adds `src/config.py` with model, training, optimizer, warmup, and augmentation settings.
- Adds `benchmarks/evaluate.py` to load YOLOv5su, train on COCO128, validate mAP50 and mAP50-95, and measure inference latency.
- Supports a `--quick` mode for latency-only screening and records results with a fixed seed.
- Creates the `artifacts` directory as needed and stores evaluation metrics in JSON format.

### 🎯 Purpose & Impact
Provides a reproducible local workflow for evaluating the proposed COCO128 fine-tuning configuration and checking inference latency; no existing model runtime behavior is changed unless this new workflow is executed.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @coderXcode, thank you for submitting a `ultralytics/yolov5` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/yolov5/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/yolov5` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Comment by coderXcode** [comment]
Opened by mistake against upstream while testing a local benchmarking tool — closing immediately. Apologies for the noise.

File: benchmarks/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            data="coco128.yaml",
+            epochs=EPOCHS,
+            batch=BATCH_SIZE,
+            imgsz=IMG_SIZE,
+            lr0=LR0,
+            lrf=LRF,
+            momentum=MOMENTUM,
+            weight_decay=WEIGHT_DECAY,
+            warmup_epochs=WARMUP_EPOCHS,
+            optimizer=OPTIMIZER,
```
Comment (UltralyticsAssistant) [inline benchmarks/evaluate.py:98]: 💡 **MEDIUM**: `OPTIMIZER` defaults to `"auto"`, under which Ultralytics ignores the supplied `lr0` and `momentum` values. Consequently, changing `LR0` or `MOMENTUM` in `src/config.py` has no effect on this experiment, so the reported run is not using the configured hyperparameters. Use an explicit optimizer when these values are intended to be tuned, or remove the inactive settings.

File: benchmarks/evaluate.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            "secondary_metrics": {
+                "latency_ms": round(latency_ms, 2),
+            },
+            "seed": 42,
+        }
+    else:
+        print(f"Starting training on COCO128 for {EPOCHS} epochs...")
+        train_start = time.time()
+        model.train(
+            data="coco128.yaml",
```
Comment (UltralyticsAssistant) [inline benchmarks/evaluate.py:89]: 📝 **LOW**: The dataset is passed as a bare relative filename, so invoking `python /path/to/benchmarks/evaluate.py` from outside the repository can resolve a different `coco128.yaml` or fail before training instead of using this checkout's `data/coco128.yaml`. Resolve the path from `__file__` and reuse it for both training and validation.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the new evaluation harness and configuration. The core training/validation flow is otherwise straightforward, but the dataset path is not anchored to the checkout and the configured learning-rate/momentum settings are inactive with the default optimizer. These should be corrected before merging.

<details><summary>💬 Posted 2 inline comments</summary>

- 💡 **MEDIUM** `benchmarks/evaluate.py:98` `OPTIMIZER` defaults to `"auto"`, under which Ultralytics ignores the supplied `lr0` and `momentum` values. Consequently, changing `LR0` or `MOMENTUM` in `src/config.py` has no effect on this experiment, so the reported run is not using the configured hyperparameters. Use an explicit optimizer when these values are intended to be tuned, or remove the inactive settings.
- 📝 **LOW** `benchmarks/evaluate.py:89` The dataset is passed as a bare relative filename, so invoking `python /path/to/benchmarks/evaluate.py` from outside the repository can resolve a different `coco128.yaml` or fail before training instead of using this checkout's `data/coco128.yaml`. Resolve the path from `__file__` and reuse it for both training and validation.
</details>
